frontend/src/pages/RestorePassword.tsx

```diff
@@ -0,0 +1,142 @@
+import React, { useState, useEffect } from "react";
+import { useSearchParams, useNavigate } from "react-router-dom";
+
+type Lang = "uk" | "en";
+type Props = { lang?: Lang };
+
+const TR = {
+  uk: {
+    title: "Встановлення нового пароля",
+    newPassword: "Новий пароль",
+    confirmPassword: "Підтвердьте пароль",
+    submit: "Змінити пароль",
+    required: "Обидва поля обов’язкові.",
+    mismatch: "Паролі не збігаються.",
+    tooShort: "Пароль має містити щонайменше 8 символів.",
+    success: "Ваш пароль успішно змінено. Зараз ви будете перенаправлені на сторінку входу.",
+    invalidToken: "Посилання недійсне або прострочене.",
+    weak: "Пароль занадто простий.",
+  },
+  en: {
+    title: "Set a new password",
+    newPassword: "New password",
+    confirmPassword: "Confirm password",
+    submit: "Change password",
+    required: "Both fields are required.",
+    mismatch: "Passwords do not match.",
+    tooShort: "Password must be at least 8 characters long.",
+    success: "Your password has been successfully changed. Redirecting to login...",
+    invalidToken: "Invalid or expired link.",
+    weak: "Password is too weak.",
+  },
+} as const;
+
+export default function RestorePassword({ lang = "uk" }: Props) {
+  const t = TR[lang];
+  const [search] = useSearchParams();
+  const navigate = useNavigate();
+
+  const [password, setPassword] = useState("");
+  const [confirm, setConfirm] = useState("");
+  const [msg, setMsg] = useState<string | null>(null);
+  const [error, setError] = useState<string | null>(null);
+  const [loading, setLoading] = useState(false);
+
+  const uid = search.get("uid");
+  const token = search.get("token");
+
+  useEffect(() => {
+    if (!uid || !token) {
+      setError(t.invalidToken);
+    }
+  }, [uid, token]);
+
+  async function onSubmit(e: React.FormEvent) {
+    e.preventDefault();
+    setMsg(null);
+    setError(null);
+
+    if (!password || !confirm) return setError(t.required);
+    if (password.length < 8) return setError(t.tooShort);
+    if (password !== confirm) return setError(t.mismatch);
+
```

Comment:
The component sets an error in useEffect when uid or token are missing, but onSubmit still proceeds to call the API using possibly null uid/token values (body: JSON.stringify({ uid, token, new_password: password })). This can send malformed requests or leak undefined values to the backend.

Explicitly verify uid and token inside onSubmit and prevent the request if missing. Example:
```
if (!uid || !token) {
  setError(t.invalidToken);
  return;
}
```
Place this check before setLoading(true).